### DISCUSSION TITLE
accessing headers in chapters via lua filter

### DISCUSSION THREAD

**Discussion by eeholmes** [quarto-dev/quarto-cli#10548] [Q&A]
### Description

Quarto version 1.5.55

I am writing a lua filter to edit chapter headers in a Quarto book. However my filter doesn't access the level 1 headers in chapters, only in the index.qmd.  I can see the problem in the page source but I don't know how to solve it.

`_quarto.yml`
```
project:
  type: book
book:
  title: "test"
  chapters:
    - index.qmd
    - one.qmd
filters:
  - test.lua
```

`test.lua`
```
function Header(el)
  quarto.log.output(pandoc.utils.stringify(el.level))
end
```

`index.qmd` and `one.qmd`
```
# One

## Two

### Three
```

What I see in the pandoc output. The numbers are the level of the headers encountered. My filter is not "seeing" the level 1 header in the first chapter (`one.qmd`).
```
[1/2] index.qmd
1
2
3
[2/2] one.qmd
2
3
```

What I expect to see. Both qmd are identical with the same three headers.
```
[1/2] index.qmd
1
2
3
[2/2] one.qmd
1
2
3
```

Looking at the page source, I see the problem I think. The `<h1>` with the chapter title is inside a `<header>` block. But I don't know how to change my filter to get "inside" the header block. I don't actually care about the h1 headers. I am trying to change the `<span  class="chapter-number">` that is in `<header>` but I don't know how to access it.

Here is the page source from `one.html`
```
<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">One</span></h1>
</div>
<div class="quarto-title-meta">
 </div>
</header>
<section id="two" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="two"><span class="header-section-number">2.1</span> Two</h2>
<section id="three" class="level3" data-number="2.1.1">
<h3 data-number="2.1.1" class="anchored" data-anchor-id="three"><span class="header-section-number">2.1.1</span> Three</h3>
```

Here is the page source from `index.html`
```
<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">test</h1>
</div>
<div class="quarto-title-meta"> </div>
</header>
<section id="one" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> One</h1>
<section id="two" class="level2" data-number="1.1">
<h2 data-number="1.1" class="anchored" data-anchor-id="two"><span class="header-section-number">1.1</span> Two</h2>
<section id="three" class="level3" data-number="1.1.1">
<h3 data-number="1.1.1" class="anchored" data-anchor-id="three"><span class="header-section-number">1.1.1</span> Three</h3>
```

**Comment by mcanouil** [comment]
- https://github.com/quarto-dev/quarto-cli/discussions/8999

**Reply by cscheid** [reply]
More specifically, level 1 headers are often interpreted as chapter titles, and so there's an inherent ambiguity in .qmd documents about where they could be coming from: metadata or markdown. It is admittedly a bit unfortunate that we don't have a clean way for Lua filters to access either of them.

**Reply by eeholmes** [reply]
@mcanouil Yes I saw #8999, but that is not what I am trying to do. I don't want the title. I need the chapter-number and I want to alter the <span> within the header.

**Reply by mcanouil** [reply]
I don't think you can access that via Lua.

**Reply by eeholmes** [reply]
I am able to access it via something like but it is not quite what I want since I only want the span insert on the span in that <header>
```
function Span(a)
  if a.classes[1] ~= "chapter-number" then return end
  chapnum = '<script>MathJax.config.section=' .. pandoc.utils.stringify(a.content[1]) .. '</script>'
  a.content = pandoc.RawInline('html', chapnum .. pandoc.utils.stringify(a.content[1]))
  return a
end
```

**Reply by cscheid** [reply]
> I don't think you can access that via Lua.

Yes, I think that's right. The problem is that Quarto itself doesn't actually know the chapter number at the time that the filter runs (since it depends potentially on all the other files).

**Reply by mcanouil** [reply]
Depending on the goal, JavaScript might be the answer here (possibly in combination with a LUA filter).

**Reply by eeholmes** [reply]
@cscheid I definitely can access the chapter-number. I have the function above that does 80% of what I need. It works and inserts the chapter number where I want it. Problem is that it is inserting it everywhere that span with class='chapter-number' appears and I want it only to appear int the `<header id="title-block-header">` tag.

Sorry, my minimal example was not really showing my problem but showing a simpler problem that I thought could help me solve my problem.

**Reply by cscheid** [reply]
> @cscheid I definitely can access the chapter-number.

:eyes:

TIL! (Quarto is _big_...). I think this might be easier on us if you can give us a tiny full repo with the problem.

**Reply by eeholmes** [reply]
Here is my minimal repo for what I am doing. It very nearly works https://github.com/nmfs-opensci/chapnum-equations

**Reply by eeholmes** [reply]
Ok, I *think* I might know what's going on. I think this is a pandoc.MetaBlock and I need to tell the filter to work on that. Right now it is working blocks and pandoc treats titles differently, as MetaBlocks even when in a Quarto book.

**Reply by mcanouil** [reply]
Just in case, are you using `format: native`? This format will output the AST.

**Reply by eeholmes** [reply]
I tried using `format: native` but it said it is not available for project type `book`. Sadly.

**Reply by ute** [reply]
@eeholmes , have you found out how to get the chapter number in an easy way? 
I needed this for [custom-numbered-blocks](https://github.com/ute/custom-numbered-blocks), and it was quite epic at that stage (still quarto 1.3 then). Information about chapters is contained in `meta.book.render`. You need to know the filename to identify the chapter number.
